src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts

```diff
@@ -0,0 +1,777 @@
+import { TypedArrayBufferView, range, unreachable } from '../../../../common/util/util.js';
+import { InterpolationSampling, InterpolationType } from '../../../constants.js';
+import { GPUTest } from '../../../gpu_test.js';
+import { checkElementsPassPredicate } from '../../../util/check_contents.js';
+
+export const g = makeTestGroup(GPUTest);
+
+const s_deviceToPipelineMap = new WeakMap<
+  GPUDevice,
+  {
+    texture_2d?: GPUComputePipeline;
+    texture_multisampled_2d?: GPUComputePipeline;
+  }
+>();
+
+/**
+ * Returns an object of pipelines associated
+ * by weakmap to a device so we can cache pipelines.
+ */
+function getPipelinesForDevice(device: GPUDevice) {
+  let pipelines = s_deviceToPipelineMap.get(device);
+  if (!pipelines) {
+    pipelines = {};
+    s_deviceToPipelineMap.set(device, pipelines);
+  }
+  return pipelines;
+}
+
+/**
+ * Gets a compute pipeline that will copy the given texture if passed
+ * a dispatch size of texture.width, texture.height
+ * @param device a device
+ * @param texture texture the pipeline is needed for.
+ * @returns A GPUComputePipeline
+ */
+function getCopyMultisamplePipelineForDevice(device: GPUDevice, texture: GPUTexture) {
+  const pipelineType = texture.sampleCount > 1 ? 'texture_multisampled_2d' : 'texture_2d';
+  const pipelines = getPipelinesForDevice(device);
+  let pipeline = pipelines[pipelineType];
+  if (!pipeline) {
+    const isMultisampled = pipelineType === 'texture_multisampled_2d';
+    const numSamples = isMultisampled ? 'textureNumSamples(texture)' : '1u';
+    const sampleIndex = isMultisampled ? 'sampleIndex' : '0';
+    const module = device.createShaderModule({
+      code: `
+        @group(0) @binding(0) var texture: ${pipelineType}<f32>;
+        @group(0) @binding(1) var<storage, read_write> buffer: array<f32>;
+
+        @compute @workgroup_size(1) fn cs(@builtin(global_invocation_id) id: vec3u) {
+          let numSamples = ${numSamples};
+          let dimensions = textureDimensions(texture);
+          let sampleIndex = id.x % numSamples;
+          let tx = id.x / numSamples;
+          let offset = (id.y * dimensions.x + tx) * numSamples + sampleIndex;
+          let v = vec4u(textureLoad(texture, vec2u(tx, id.y), ${sampleIndex}) * 255.0);
+
+          buffer[offset] = bitcast<f32>(dot(v, vec4u(0x1000000, 0x10000, 0x100, 0x1)));
+        }
+      `,
+    });
+
+    pipeline = device.createComputePipeline({
+      label: 'copy multisampled texture pipeline',
+      layout: 'auto',
+      compute: {
+        module,
+        entryPoint: 'cs',
+      },
+    });
+
+    pipelines[pipelineType] = pipeline;
+  }
+  return pipeline;
+}
+
+/**
+ * Copies a texture (even if multisampled) to a buffer
+ * @param t a gpu test
+ * @param texture texture to copy
+ * @returns buffer with copy of texture, mip level 0, array layer 0.
+ */
+function copyTextureToBufferIncludingMultisampledTextures(t: GPUTest, texture: GPUTexture) {
```

Comment:
That's a good idea.

For the copy, checking the code over at gpu_test `copySinglePixelTextureToBufferUsingComputePass` is (a) only copying a single pixel and (b) isn't recovering f32 values from an rgba8unorm texture. Do we want to try to combine all of those options into a single function or just keep them separate.

I agree that using texelviews to do the comparison would be really good to do. I think I'll mark it as a MAINTANENCE_TODO and try to change it in the next PR.

Reply:
I think that we should probably keep the value recovering part separate. (Might be able to do it via something internal to the `TexelViews` where we could specify how to interpret the binary data.)

But copying the data in/out would be nice to unify.

Reply:
So for now I changed it. Instead of getting the values from the textures one texture at a time and then combining them after, it reads all 4 textures at once into a single buffer. This results in a buffer that's ready to use as `rgba32float` which makes it easy to call the TexelView code. So the code now uses TexelView

We can decide whether to switch to reading them as rgba8unorm and keeping them as rgba8unorm in a separate PR, but, at a glance, the TexelView code assumes a single buffer and so, we'd either have to merge/interleave the 4 rgba8uorm buffers into one buffer or else refactor the TexelView code so it doesn't assume one buffer. As it is, it calls "theOneBuffer.subarray" to get data for each texel but that won't work here unles the buffers are interleaved into a single buffer.

There's also the issue that the TexelView code has no concept of samples. So, if it prints an error in texel 11, 5 and your sampleCount is 4 then that really means texel 2, 5, sample index 3,

It might be simple to pass in a sampleCount and have the TexelView code do `texelX = x / sampleCount | 0; sampleIndex = texelX % sampleCount` but that should probably happen in another PR

Reply:
Gotcha! This is great already! I think that copying the data out into a single buffer is fine, we probably just need to introduce the concept of samples to the TexelView. If we default it to 1, and then allow users to specify, we can just assume that the textures will be one after another in the buffer. (As long as we define and use the same copy functions we should be fine.) And yes, absolutely those follow ups should be in a separate PR! Happy holidays!!